### PR TITLE
DataGrid should apply columns that are dynamically added to a band (T815945)

### PR DESCRIPTION
**Pull request by igormalzew**

### CODE REVIEW

File: js/ui/grid_core/ui.grid_core.columns_controller.js [hunk 1]
```diff
@@ -1577,9 +1577,17 @@ module.exports = {
...
                             args.handled = true;
-                            if(!this._skipProcessingColumnsChange) {
+
+                            const endsWith = (str, subStr) => {
+                                let endIndex = str.length - subStr.length;
+                                return str.lastIndexOf(subStr) === endIndex;
+                            };
+
+                            if(args.name !== args.fullName && endsWith(args.fullName, "columns")) {
+                                setTimeout(() => this.reinit());
```
Comment (AntonSermyazhko) [inline js/ui/grid_core/ui.grid_core.columns_controller.js]: Can we get rid of setTimeout here?

File: testing/tests/DevExpress.ui.widgets.dataGrid/dataGrid.tests.js [hunk 2]
```diff
@@ -4434,6 +4434,39 @@ QUnit.test("Freespace row have the correct height when using master-detail with
     this.clock.restore();
 });
 
+QUnit.test("DataGrid should apply columns that are dynamically added to a band (T815945)", function(assert) {
+    // arrange
+    this.clock = sinon.useFakeTimers();
```
Comment (AntonSermyazhko) [inline testing/tests/DevExpress.ui.widgets.dataGrid/dataGrid.tests.js]: I suggest you use a local variable to initialize fake timers. Also the module can already has an initialized this.clock. If so, you can use it.

Comment (igormalzew) [reply testing/tests/DevExpress.ui.widgets.dataGrid/dataGrid.tests.js]: The module has not an initialized this.clock

**Review by AntonSermyazhko — DISMISSED** [review]

File: js/ui/grid_core/ui.grid_core.columns_controller.js [hunk 3]
```diff
@@ -1213,7 +1213,9 @@ module.exports = {
 
             var fireColumnsChanged = function(that) {
                 var onColumnsChanging = that.option("onColumnsChanging"),
-                    columnChanges = that._columnChanges;
+                    columnChanges = that._columnChanges,
+                    reinitOptionNames = ["dataField", "lookup", "dataType", "columns"],
+                    needReinit = (options) => options && reinitOptionNames.filter(name => options[name]).length > 0;
```
Comment (vconst) [inline js/ui/grid_core/ui.grid_core.columns_controller.js]: filter -> some

**Review by vconst — CHANGES_REQUESTED** [review]

**Review by AntonSermyazhko — APPROVED** [review]

**Review by AntonSermyazhko — APPROVED** [review]

**Review by vconst — APPROVED** [review]

**Review by jtoming830 — APPROVED** [review]

**Pull request merged**